ServantPersistent/src/Domain/Product.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Domain.Product where
+
+import           Control.Monad.Except
+import           Data.Maybe
+import qualified Data.Text            as T
+import           Data.Time
+import           Database.Persist
+import           DBTypes
+import           Models
+import           Operation
+import           ProductQuery
+import           System.Random
+import           Types
+
+data ProductCreationError = EmptyVariantList
+                          | PhysicalProductFieldsMissing
+                          | DigitalProductExtraFields
+                          | ProductUniquenessViolated
+
+sluggify :: T.Text -> T.Text
+sluggify = T.intercalate "-"
+         . fmap T.toLower
+         . T.words
+
+makeUnique :: MonadIO m => T.Text -> TransactionT m T.Text
+makeUnique s = do
+  r <- getBy (UniqueSlug s)
+  case r of
+    Nothing -> return s
+    Just _ -> do
+           (n :: Int) <- liftIO $ randomRIO (1000,100000)
+           makeUnique (s `T.append` T.pack (show n))
+
+dbCreateProduct :: MonadIO m
+                => TenantId
+                -> ProductInput
+                -> OperationT (TransactionT m) (Either ProductCreationError ProductId)
+dbCreateProduct tid ProductI{..} =
+  requirePermission (CreateProduct tid) $
+  lift $ runExceptT $ do
+     time <- liftIO getCurrentTime
+     when (null piVariants) $
+          throwError EmptyVariantList
+     case piType of
```

Comment:
Possible to avoid this runtime check by making illegal states unrepresentable in our domain?